js/src/path.js

```diff
@@ -4,43 +4,220 @@
 // Licensed under the Apache License, version 2.0:
 // http://www.apache.org/licenses/LICENSE-2.0
 
+import {invariant} from './assert.js';
 import type Value from './value.js';
+import Hash from './hash.js';
 import {Kind} from './noms-kind.js';
 import List from './list.js';
 import Map from './map.js';
+import Set from './set.js';
+import {OrderedKey} from './meta-sequence.js';
+import {OrderedSequence} from './ordered-sequence.js';
+import {fieldNameComponentRe} from './struct.js';
 import {getTypeOfValue, StructDesc} from './type.js';
 
-interface Part {
-  resolve(v: Promise<?Value>): Promise<?Value>;
+const annotationRe = /^@([a-z]+)/;
+
+export interface Part {
+  resolve(v: Value): Promise<?Value>;
   toString(): string;
 }
 
-class FieldPart {
-  name: string;
+/**
+ * A Path is an address to a Noms value - and unlike hashes (i.e. #abcd...) they can address inlined
+ * values. See https://github.com/attic-labs/noms/blob/master/doc/spelling.md.
+ */
+export default class Path {
+  _parts: Array<Part>;
 
-  constructor(name: string) {
-    this.name = name;
+  static parse(str: string): Path {
+    if (str === '') {
+      throw new SyntaxError('Empty path');
+    }
+    const p = new Path();
+    p._parse(str);
+    return p;
   }
 
-  resolve(v: Promise<?Value>): Promise<?Value> {
-    return v.then(value => {
-      if (value === null || value === undefined) {
-        return;
+  constructor(parts: Array<Part> = []) {
+    this._parts = parts;
+  }
+
+  append(part: Part): Path {
+    return new Path(this._parts.concat(part));
+  }
+
+  toString(): string {
+    return this._parts.map(p => p.toString()).join('');
+  }
+
+  async resolve(v: Value): Promise<?Value> {
+    let res = v;
+    for (const part of this._parts) {
+      if (res == null) {
```

Comment:
Use undefined. Testing for both is a sign that we cannot decide which one we want to use.